presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java

```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
 
         PlanNode root = subPlan.getRoot();
         if (root.getOutputSymbols().isEmpty()) {
             // there is nothing to join with - e.g. SELECT (SELECT 1)
-            return new PlanBuilder(translations, enforceSingleRowNode, subPlan.getSampleWeight());
+            return new PlanBuilder(translations, subquery, subPlan.getSampleWeight());
         }
         else {
             return new PlanBuilder(translations,
-                    // TODO handle parameter list
                     new ApplyNode(idAllocator.getNextId(),
                             root,
-                            enforceSingleRowNode,
-                            ImmutableList.of()),
+                            subquery,
+                            ImmutableList.copyOf(correlation.values())),
                     subPlan.getSampleWeight());
         }
     }
 
+    private Map<Expression, Symbol> extractCorrelation(PlanBuilder subPlan, PlanNode subquery)
+    {
+        Set<Expression> missingReferences = extractMissingExpressions(subquery);
+        ImmutableMap.Builder<Expression, Symbol> correlation = ImmutableMap.builder();
+        if (!missingReferences.isEmpty()) {
+            for (Expression missingReference : missingReferences) {
+                Expression rewritten = subPlan.rewrite(missingReference);
+                if (rewritten != missingReference) {
+                    correlation.put(missingReference, asSymbol(rewritten));
+                }
+            }
+        }
+        return correlation.build();
+    }
+
     private RelationPlan createRelationPlan(SubqueryExpression subqueryExpression)
     {
         return new RelationPlanner(analysis, symbolAllocator, idAllocator, metadata, session)
                 .process(subqueryExpression.getQuery(), null);
     }
+
+    private Symbol asSymbol(Expression expression)
+    {
+        checkState(expression instanceof QualifiedNameReference);
+        return Symbol.fromQualifiedName(((QualifiedNameReference) expression).getName());
+    }
+
+    private Set<Expression> extractMissingExpressions(PlanNode planNode)
+    {
+        MissingReferencesExtractor visitor = new MissingReferencesExtractor(analysis);
+        planNode.accept(visitor, null);
+        return visitor.getMissing();
+    }
+
+    private static class MissingReferencesExtractor
+            extends PlanVisitor<Void, Void>
+    {
+        private final Set<Expression> missing = new HashSet<>();
+        private final Set<Expression> columnReferences;
+
+        private MissingReferencesExtractor(Analysis analysis)
+        {
+            columnReferences = analysis.getColumnReferences();
+        }
+
+        @Override
+        protected Void visitPlan(PlanNode node, Void context)
+        {
+            node.getSources().forEach(source -> source.accept(this, null));
+
+            for (Symbol symbol : node.getOutputSymbols()) {
+                missing.remove(symbol.toQualifiedNameReference());
```

Comment:
This doesn't look right. There's no relationship between the names of the outputs and the column references in the analysis.